### PR TITLE
[CELEBORN-1258] Add UserIdentifier to the master application meta

### PR DESCRIPTION
**Pull request by turboFei**
### What changes were proposed in this pull request?

In this pr, we support to update the user identifier in application meta when initializing lifecycle manager.

### Why are the changes needed?
to close CELEBORN-1285

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

UT.

### CODE REVIEW

**Comment by turboFei** [comment]
cc @RexXiong 

Could you help have a overall review? I can enhance it after that.

**Comment by RexXiong** [comment]
> cc @RexXiong
> 
> Could you help have a overall review? I can enhance it after that.

OK, I will review this asap.

File: common/src/main/scala/org/apache/celeborn/common/meta/ApplicationInfo.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package org.apache.celeborn.common.meta
+
+import java.util.concurrent.atomic.LongAdder
+
+/**
+ * Application Info.
+ */
+class ApplicationInfo {
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/meta/ApplicationInfo.scala:25]: why we need add this class? and it seems irrelevant with this pr

File: common/src/main/scala/org/apache/celeborn/common/util/PbSerDeUtils.scala [hunk 2]
```diff
@@ -457,6 +459,22 @@ object PbSerDeUtils {
...
 
+  def fromPbApplicationInfo(pbApplicationInfo: PbApplicationInfo): ApplicationInfo = {
+    val applicationInfo = new ApplicationInfo()
+    applicationInfo.updateFileCount(pbApplicationInfo.getFileCount)
+    applicationInfo.updateTotalWritten(pbApplicationInfo.getTotalWritten)
+    applicationInfo.setHeartbeatTime(pbApplicationInfo.getLastHeartbeatTime)
+    applicationInfo
+  }
+
+  def toPbApplicationInfo(applicationInfo: ApplicationInfo): PbApplicationInfo = {
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/util/PbSerDeUtils.scala:470]: PbApplicationMeta need add userIdentifier and should serialize/deserialize to/from ratis log.

**Comment by mridulm** [comment]
A general query on this - is there any benefit to not supporting auth ?
This change is not backwardly compatible - so existing applications cannot leverage it - for newer deployments and/or upgrade - make authn standard ?

**Comment by turboFei** [comment]
>is there any benefit to not supporting auth going forward given authn support has completed ?

I agree, we will also enable auth for our use case.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 3]
```diff
@@ -1142,12 +1141,26 @@ class CelebornConf(loadDefaults: Boolean) extends Cloneable with Logging with Se
     authEnabled && internalPortEnabled
   }
 
+  def appRegisterEnabled: Boolean = {
+    val appRegisterEnabled = get(APP_REGISTER_ENABLED) || authEnabled
+    val internalPortEnabled = get(INTERNAL_PORT_ENABLED)
```
[the remark this answers is not in the document]
Comment (turboFei) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1146]: without internalPort, worker can not interactive with the master rpc endpoint directly, as there is a RegisterationRpcHandler

**Comment by turboFei** [comment]
This pr is used to provide the tenant information for dashboard. https://issues.apache.org/jira/browse/CELEBORN-1227 

Do we plan to set `celeborn.auth.enabled` to true by defaults?

**Comment by RexXiong** [comment]
> This pr is used to provide the tenant information for dashboard. https://issues.apache.org/jira/browse/CELEBORN-1227
> 
> Do we plan to set `celeborn.auth.enabled` to true by defaults?

No, IMO this feature has nothing to do with whether the auth enabled or not.

**Comment by otterc** [comment]
Have a general question. Why is user identifier needed to be shared with Celeborn? I can't find any information in the linked jira: [CELEBORN-1285](https://issues.apache.org/jira/browse/CELEBORN-1285)
Also, I think if registration is required for the feature that we are trying to support, I think it will better to add a feature flag instead of `register.enabled` flag.

**Comment by turboFei** [comment]
> Have a general question. Why is user identifier needed to be shared with Celeborn? I can't find any information in the linked jira: [CELEBORN-1285](https://issues.apache.org/jira/browse/CELEBORN-1285) Also, I think if registration is required for the feature that we are trying to support, I think it will better to add a feature flag instead of `register.enabled` flag.

Thanks for the suggestion, will check how to implement it.

**Comment by turboFei** [comment]
thanks for your suggestions, could you help take a look again?

@RexXiong @mridulm @otterc

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 4]
```diff
@@ -173,7 +173,7 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
   if (authEnabled) {
     logInfo(s"Authentication is enabled; setting up master and worker RPC environments")
     val appSecret = createSecret()
-    applicationMeta = ApplicationMeta(appUniqueId, appSecret)
+    applicationAuthMeta = ApplicationMeta(appUniqueId, appSecret)
```
Comment (RexXiong) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:176]: Add else to initialize ApplicationMeta when auth disabled?

Comment (turboFei) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala]: thanks, addressed.

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 5]
```diff
@@ -113,7 +113,7 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
 
   private val mockDestroyFailure = conf.testMockDestroySlotsFailure
   private val authEnabled = conf.authEnabledOnClient
-  private var applicationMeta: ApplicationMeta = _
+  private var applicationAuthMeta: ApplicationMeta = _
```
[the remark this answers is not in the document]
Comment (turboFei) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala]: thanks, addressed

**Review by RexXiong — APPROVED** [review]
Should also deserialize the userIndentifier of  ApplicationMeta in MetaHandler.

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 6]
```diff
@@ -207,13 +209,20 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
...
+  }
+
   // Since method `onStart` is executed when `rpcEnv.setupEndpoint` is executed, and
   // `masterClient` is initialized after `rpcEnv` is initialized, if method `onStart` contains
   // a reference to `masterClient`, there may be cases where `masterClient` is null when
   // `masterClient` is called. Therefore, it's necessary to uniformly execute the initialization
   // method at the end of the construction of the class to perform the initialization operations.
   private def initialize(): Unit = {
     // noinspection ConvertExpressionToSAM
+    updateApplicationMeta()
```
Comment (otterc) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:225]: This actually conflicts with safely propagating the application secret to the Celeborn Master. When auth is enabled, this will transmit application secret to Celeborn Master without any Sasl client authentication. Currently, we have added Anonymous client authentication, but the plan was to add other mechanisms.
cc. @mridulm

Comment (otterc) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:225]: I think we should rename the existing `ApplicationMeta` to `ApplicationAuthMeta` and it should contain information specific to authentication. We can then use ApplicationMeta for general app info that needs to be sent to the Celeborn Master.

Comment (turboFei) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:225]: checking

Comment (turboFei) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:225]: raised https://github.com/apache/celeborn/pull/2420 to separate application auth and general meta. cc @otterc @RexXiong

Comment (RexXiong) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:225]: > This actually conflicts with safely propagating the application secret to the Celeborn Master. When auth is enabled, this will transmit application secret to Celeborn Master without any Sasl client authentication. Currently, we have added Anonymous client authentication, but the plan was to add other mechanisms. cc. @mridulm

Based on my understanding, if authentication is enabled in Celeborn, it is not possible to access the Celeborn Master without SASL client authentication. This PR does not introduce new mechanisms; it merely adds an identifier to ApplicationMeta. Therefore, I believe this PR does not compromise security. @otterc

Comment (otterc) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:225]: Yes, you are correct. The connection will only be established once the client is authenticated. However, I still believe that sending the secret in plain text multiple times to the Master may cause issues in the future. Eventually, we would want to encrypt the secret, and TLS support is for that purpose. If we are sending the secret multiple times, it would mean that not only registration has to be done with TLS but also that all messages should be sent via TLS. Therefore, I think having authentication metadata separated from general application metadata will be better.

Comment (RexXiong) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:225]: Agree with most messages shouldn't be sent via TLS. I believe that the most fundamental issue with the existing security authentication mechanism is that it only authenticates the connection, but does not verify the legitimacy of the messages sent by the authenticated client. At the very least, we need verify that the applicationId in the sent messages matches the applicationId provided during the initial authentication, Otherwise, an authenticated client could still access or modify with the data of other applications. I am not sure if this is in line with the expectations.

Comment (otterc) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:225]: Yes, this is a gap. I created https://issues.apache.org/jira/browse/CELEBORN-1360 to address this. Will open a PR soon.

Comment (RexXiong) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:225]: > Yes, this is a gap. We can validate the appId in the fetch and push messages of an application. I created https://issues.apache.org/jira/browse/CELEBORN-1360 for that and will create a PR soon.

If that's the case, then we won't have to send a secret; we'll only need to update the userIdentifier. I believe that extending the ApplicationMeta to contain more application-related information is acceptable.
